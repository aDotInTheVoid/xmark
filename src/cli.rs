use clap::Clap;
use eyre::Result;
use std::path::PathBuf;

#[derive(Debug, Default, Clone)]
pub struct Args {
    pub dir: PathBuf,
    pub create: bool,
    pub templates: Option<PathBuf>,
}

impl Args {
    pub fn parse() -> Result<Self> {
        let ArgsInner {
            mut dir,
            create,
            mut templates,
        } = ArgsInner::parse();
        dir = dir.canonicalize()?;
        templates = match templates {
            Some(p) => Some(p.canonicalize()?),
            None => None,
        };
        Ok(Args {
            dir,
            create,
            templates,
        })
    }
}

#[derive(Clap, Debug)]
struct ArgsInner {
    #[clap(short, long, default_value = ".")]
    pub dir: PathBuf,
    #[clap(long)]
    pub create: bool,
    #[clap(short, long)]
    pub templates: Option<PathBuf>,
}
