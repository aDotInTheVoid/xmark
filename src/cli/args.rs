use clap::Clap;
use eyre::Result;
use tracing::instrument;

#[derive(Clap, Debug)]
pub(super) struct Args {
    #[clap(subcommand)]
    subcmd: SubCommand,
}

#[derive(Clap, Debug)]
pub(super) enum SubCommand {
    Build(super::cmd::build::Args), // TODO: init, clean, watch, serve
}

impl Args {
    #[instrument]
    pub fn run(self) -> Result<()> {
        match self.subcmd {
            SubCommand::Build(c) => c.run(),
        }
    }
}
