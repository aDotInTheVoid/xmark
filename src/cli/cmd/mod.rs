//! Individual subcommands.
pub mod build;