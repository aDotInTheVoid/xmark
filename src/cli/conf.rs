//! The xmark.toml file
use serde::{Deserialize, Serialize};
use toml::{self, value::Table};

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct DiskConf {
    pub books: Vec<Location>,
    #[serde(default)]
    pub output: Table,
}
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(untagged)]
pub enum Location {
    Bare(String),
    Named { name: String, root_loc: String },
}

#[cfg(test)]
mod tests {
    use super::*;
    use toml::Value;

    fn check_disc_conf(l: DiskConf, r: &str) {
        assert_eq!(toml::from_str::<DiskConf>(r).unwrap(), l);
    }

    fn l(s: &str) -> Location {
        Location::Bare(s.to_owned())
    }

    #[test]
    fn basic_disc_conf() {
        check_disc_conf(
            DiskConf {
                books: vec![l("hello"), l("w2"), l("x4")],
                output: Table::new(),
            },
            r#"
books = [
    "hello",
    "w2",
    "x4",
]
            "#,
        );
    }

    #[test]
    fn empty_output_html() {
        check_disc_conf(
            DiskConf {
                books: vec![l("just_one")],
                output: vec![("html".to_owned(), Value::Table(Table::new()))]
                    .into_iter()
                    .collect(),
            },
            r#"
books = [ "just_one" ]
[output.html]
            "#,
        )
    }

    #[test]
    fn output_html_vals() {
        let mut html = Table::new();
        html.insert(
            "output-url".to_owned(),
            Value::String("/bookshelf/".to_owned()),
        );
        html.insert("minify".to_owned(), Value::Boolean(false));
        let html = Value::Table(html);

        check_disc_conf(
            DiskConf {
                books: vec![l("the-book")],
                output: vec![("html".to_owned(), html)].into_iter().collect(),
            },
            r#"
books = [ "the-book" ]
[output.html]
output-url = "/bookshelf/"
minify = false
            "#,
        )
    }

    #[test]
    fn multi_output() {
        let mut html = Table::new();
        html.insert(
            "output-url".to_owned(),
            Value::String("/bookshelf/".to_owned()),
        );
        html.insert("minify".to_owned(), Value::Boolean(false));
        let html = Value::Table(html);

        let mut tex = Table::new();
        tex.insert("created".to_owned(), Value::Integer(1978));
        tex.insert("xml".to_owned(), Value::Boolean(false));
        tex.insert("parsing".to_owned(), Value::String("Imposible".to_owned()));
        let tex = Value::Table(tex);

        check_disc_conf(
            DiskConf {
                books: Vec::new(),
                output: vec![("html".to_owned(), html), ("tex".to_owned(), tex)]
                    .into_iter()
                    .collect(),
            },
            r#"
books = []
[output.tex]
created = 1978
xml = false
parsing = "Imposible"
# Comment
[output.html]
output-url = "/bookshelf/"
minify = false
        "#,
        )
    }

    #[test]
    fn with_loc() {
        check_disc_conf(
            DiskConf {
                books: vec![
                    l("bare"),
                    Location::Named {
                        name: "trpl".to_owned(),
                        root_loc: "book/src/".to_owned(),
                    },
                    l("another-bare"),
                    Location::Named {
                        name: "cxx".to_owned(),
                        root_loc: "cxx/guide/src".to_owned(),
                    },
                    l("xmark"),
                    l("dummy"),
                ],
                output: Table::new(),
            },
            r#"
books = [
    "bare",
    { location = "book/src/", name = "trpl" },
    "another-bare",
    { name = "cxx", location="cxx/guide/src" },
    "xmark",
    "dummy",
]
[output]
        "#,
        )
    }
}
