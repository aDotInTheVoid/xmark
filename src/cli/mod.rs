// SPDX-License-Identifier: GPL-3.0-only
use clap::Clap;
use eyre::Result;
use serde::{Deserialize, Serialize};
use std::path::PathBuf;
use tracing::instrument;

pub mod config;
pub mod summary;

mod create_missing;

#[derive(Debug, Default, Clone, Serialize, Deserialize)]
pub struct Args {
    pub dir: PathBuf,
    pub create: bool,
    pub templates: Option<PathBuf>,
}

impl Args {
    #[instrument]
    pub fn parse() -> Result<Self> {
        Self::parse_from(ArgsInner::parse())
    }

    #[instrument]
    fn parse_from(inner: ArgsInner) -> Result<Self> {
        let ArgsInner {
            mut dir,
            create,
            mut templates,
        } = inner;
        dir = dir.canonicalize()?;
        templates = match templates {
            Some(p) => Some(p.canonicalize()?),
            None => None,
        };
        Ok(Args {
            dir,
            create,
            templates,
        })
    }
}

#[instrument]
pub fn init() -> Result<()> {
    color_eyre::install()?;

    use tracing_error::ErrorLayer;
    use tracing_subscriber::prelude::*;
    use tracing_subscriber::{fmt, EnvFilter};

    let fmt_layer = fmt::layer().with_target(false);
    let filter_layer = EnvFilter::try_from_default_env()
        .or_else(|_| EnvFilter::try_new("info"))
        .unwrap();

    tracing_subscriber::registry()
        .with(filter_layer)
        .with(fmt_layer)
        .with(ErrorLayer::default())
        .init();

    Ok(())
}

#[derive(Clap, Debug)]
struct ArgsInner {
    #[clap(short, long, default_value = ".")]
    pub dir: PathBuf,
    #[clap(long)]
    pub create: bool,
    #[clap(short, long)]
    pub templates: Option<PathBuf>,
}

#[cfg(test)]
mod tests {
    use insta::{assert_yaml_snapshot, dynamic_redaction};

    use crate::test_utils::manifest_dir_redacter;

    use super::*;
    #[test]
    fn canonicalize_cwd_dir() {
        let args = ArgsInner {
            dir: ".".into(),
            create: true,
            templates: None,
        };
        let args = Args::parse_from(args).unwrap();
        // Yaml uses ~ for null
        assert_yaml_snapshot!(args, {
            ".*" => dynamic_redaction(manifest_dir_redacter)
        });
    }

    #[test]
    fn canonicalize_rel_templ() {
        let args = ArgsInner {
            dir: ".".into(),
            create: true,
            // This needs to exist, as `.canonicalize` will read, in case it's a symblink
            templates: Some("./www".into()),
        };
        let args = Args::parse_from(args).unwrap();
        assert_yaml_snapshot!(args, {
            ".*" => dynamic_redaction(manifest_dir_redacter)
        });
    }

    #[test]
    fn absolute() {
        let args = ArgsInner {
            // This should exist on most non windows systems
            // TODO: win
            dir: "/usr".into(),
            create: false,
            templates: None,
        };
        let args = Args::parse_from(args).unwrap();
        assert_yaml_snapshot!(args, {
            ".*" => dynamic_redaction(manifest_dir_redacter)
        });
    }
}
