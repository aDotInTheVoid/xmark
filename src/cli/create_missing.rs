// SPDX-License-Identifier: GPL-3.0-only
use super::summary::{Link, Summary};
use eyre::Result;
use std::fs::{self, File};
use std::io::Write;
use std::path::Path;

use log::debug;

pub fn create_missing(src_dir: &Path, summary: &Summary) -> Result<()> {
    let mut items: Vec<_> = summary
        .prefix_chapters
        .iter()
        .chain(summary.suffix_chapters.iter())
        .map(|chapter| Link {
            // This is like 3 strings
            chapter: chapter.clone(),
            ..Default::default()
        })
        //TODO: Don't clone, it can be quite nested
        .chain(summary.numbered_chapters.iter().cloned())
        .collect();

    while !items.is_empty() {
        let link = items.pop().expect("already checked");

        if let Some(ref location) = link.chapter.location {
            let filename = src_dir.join(location);
            if !filename.exists() {
                if let Some(parent) = filename.parent() {
                    if !parent.exists() {
                        fs::create_dir_all(parent)?;
                    }
                }
                debug!("Creating missing file {}", filename.display());

                let mut f = File::create(&filename)?;
                writeln!(f, "# {}", link.chapter.name)?;
            }
        }

        items.extend(link.nested_items);
    }

    Ok(())
}

#[cfg(test)]
mod tests {
    use std::collections::BTreeSet;
    use std::path::PathBuf;

    use assert_fs::prelude::*;
    use insta::{assert_yaml_snapshot, dynamic_redaction};

    use crate::cli::{self, config};
    use crate::html_render;

    #[test]
    fn dummy_e2e() {
        let temp = assert_fs::TempDir::new().unwrap();

        temp.copy_from(
            PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("dummy-book"),
            &["xmark.toml", "*/SUMMARY.md"],
        )
        .unwrap();

        let args = cli::Args {
            dir: temp.path().to_owned(),
            create: true,
            ..Default::default()
        };

        let conf = config::load(&args).unwrap();
        let render = html_render::HTMLRender::new(&conf, &args).unwrap();
        render.render().unwrap();

        // BTree so it's in order.
        let paths: BTreeSet<_> = ignore::Walk::new(temp.path())
            .filter_map(|x| x.ok().map(|y| y.into_path()))
            .map(|x| x.into_os_string().into_string().unwrap())
            .filter(|x| !x.contains("_out"))
            .collect();

        assert_yaml_snapshot!(paths, {"[]" => dynamic_redaction(move |val, _| {
            val.as_str().unwrap().replace(temp.path().as_os_str().to_str().unwrap(), "")
        })});
    }
}
