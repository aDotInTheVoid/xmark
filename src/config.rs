use std::fs;
use std::path::PathBuf;

use eyre::{Result, WrapErr};
use serde::{Deserialize, Serialize};

use crate::cli;
use crate::create_missing::create_missing;
use crate::summary::{parse_summary, Chapter, Summary};

/// The Config as represented in the global xmark.toml
#[derive(Clone, Debug, Hash, Serialize, Deserialize)]
pub struct GlobalConfigRepr {
    pub books: Vec<String>,
}

/// The config as usable for the programm
#[derive(Clone, Debug, Hash, PartialEq, Serialize, Deserialize)]
pub struct GlobalConf {
    pub books: Vec<BookConf>,
}

#[derive(Clone, Debug, Hash, PartialEq, Serialize, Deserialize)]
pub struct BookConf {
    pub location: PathBuf,
    pub slug: String,
    pub summary: Summary,
}

pub fn load(args: &cli::Args) -> Result<GlobalConf> {
    let conf = fs::read_to_string(args.dir.clone().join("xmark.toml"))
        .with_context(|| "Couldn't find xmark.toml")?;
    let conf: GlobalConfigRepr = toml::from_str(&conf)?;

    hydrate(conf, &args)
}

// Convert the disk format to a usable form
pub fn hydrate(gcr: GlobalConfigRepr, args: &cli::Args) -> Result<GlobalConf> {
    Ok(GlobalConf {
        books: gcr
            .books
            .iter()
            .map(|name| {
                let location = args.dir.join(name);
                let summary_location = location.join("SUMMARY.md");
                let mut summary = parse_summary(
                    &fs::read_to_string(&summary_location)
                        .wrap_err_with(|| format!("Couldn't open {:?}", summary_location))?,
                )?;

                if args.create {
                    create_missing(&location, &summary)?;
                }

                let fix_chap_loc = |chap: &mut Chapter| {
                    if let Some(loc) = chap.location.as_deref() {
                        chap.location = Some(location.join(loc));
                    }
                };

                summary.prefix_chapters.iter_mut().for_each(fix_chap_loc);

                summary.suffix_chapters.iter_mut().for_each(fix_chap_loc);

                summary.numbered_chapters.iter_mut().for_each(|chap| {
                    chap.map_mut(fix_chap_loc);
                });

                Ok(BookConf {
                    location,
                    summary,
                    slug: name.to_owned(),
                })
            })
            .collect::<Result<_>>()?,
    })
}

#[cfg(test)]
mod tests {
    use insta::assert_yaml_snapshot;

    use super::*;

    #[test]
    fn de_gloal_config() {
        let inp = "books = [
            'book-1',
            'book-2',
            'book-3',
            ]
";
        let conf: GlobalConfigRepr = toml::from_str(inp).unwrap();
        assert_eq!(conf.books, vec!["book-1", "book-2", "book-3"]);
    }

    #[test]
    fn hydrate_basic() {
        let args = cli::Args {
            dir: "/home/etc/bax".into(),
            create: false,
        };
        let gcr = GlobalConfigRepr { books: vec![] };
        let gc = GlobalConf { books: vec![] };
        assert_eq!(hydrate(gcr, &args).unwrap(), gc);
    }

    #[test]
    fn hydrate_dummy() {
        // I don't know why I need the anotation
        let redacter = |mut val, _: insta::internals::ContentPath| {
            // TODO: Dont do insta crimes.
            // This is in #[doc(Hidden)] internals.
            while let insta::internals::Content::Some(some) = val {
                val = *some;
            }

            if let insta::internals::Content::String(s) = val {
                s.replace(dbg!(env!("CARGO_MANIFEST_DIR")), "BASEDIR")
            } else {
                "".into()
            }
        };

        let dir = PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("dummy-book");
        let args = cli::Args { dir, create: false };
        let conf = load(&args).unwrap();
        assert_yaml_snapshot!(conf, {
            ".location" => insta::dynamic_redaction(redacter),
            ".**.location" => insta::dynamic_redaction(redacter),
        });
    }
}
