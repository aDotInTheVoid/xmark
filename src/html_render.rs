use std::fs;
use std::io::Write;
use std::path::Path;

use eyre::{Context, Result};
use pulldown_cmark::{html, Options, Parser};

use crate::cli;
use crate::config::BookConf;
use crate::summary::Chapter;

pub fn render(book: BookConf, args: &cli::Args) -> Result<()> {
    let out_dir = args.dir.clone().join("_out").join("html");
    fs::create_dir_all(&out_dir)?;

    for i in book.summary.prefix_chapters {
        render_chap_io(&i, &out_dir, &args.dir)?;
    }
    for i in book.summary.suffix_chapters {
        render_chap_io(&i, &out_dir, &args.dir)?;
    }
    for i in book.summary.numbered_chapters {
        i.try_map(|chap| render_chap_io(chap, &out_dir, &args.dir))?;
    }

    Ok(())
}

fn render_chap_io(chapter: &Chapter, build_dir: &Path, base_dir: &Path) -> Result<()> {
    // Not a draft
    if let Some(ref loc) = chapter.location {
        let content = fs::read_to_string(loc)?;
        let html = render_chap(&content);

        let mut path = build_dir.join(loc.strip_prefix(base_dir).expect("Unreachble"));

        path.set_extension("html");

        fs::create_dir_all(path.parent().unwrap())?;

        let mut file =
            fs::File::create(&path).wrap_err_with(|| format!("Failed to create {:?}", &path))?;
        file.write_all(html.as_bytes())?;
    }

    Ok(())
}

// Pure inner for testing
fn render_chap(content: &str) -> String {
    let opts = Options::all();
    let parser = Parser::new_ext(content, opts);
    let mut out = String::new();
    html::push_html(&mut out, parser);
    out
}

#[cfg(test)]
mod tests {
    use std::collections::{BTreeSet};
    use std::path::PathBuf;

    use assert_fs::prelude::*;
    use insta::{assert_yaml_snapshot, dynamic_redaction};

    use crate::{cli, config, html_render};

    #[test]
    fn dummy_e2e() {
        let temp = assert_fs::TempDir::new().unwrap();
        temp.copy_from(
            PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("dummy-book"),
            &["xmark.toml", "book-*/**"],
        )
        .unwrap();

        let args = cli::Args {
            create: false,
            dir: temp.path().to_owned(),
        };
        let conf = config::load(&args).unwrap();
        for book in conf.books {
            html_render::render(book, &args).unwrap()
        }


        // BTree so it's in order.
        let paths: BTreeSet<_> = ignore::Walk::new(temp.path())
            .filter_map(|x| x.ok().map(|y| y.into_path()))
            .map(|x| x.into_os_string().into_string().unwrap())
            .collect();

        assert_yaml_snapshot!(paths, {"[]" => dynamic_redaction(move |val, _| {
            val.as_str().unwrap().replace(temp.path().as_os_str().to_str().unwrap(), "")
        })});
    }
}
