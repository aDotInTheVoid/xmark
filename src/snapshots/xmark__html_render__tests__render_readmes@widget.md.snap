---
source: src/html_render.rs
expression: out
input_file: src/render_html_tests/widget.md
---
<h1>Judge0 Widgets</h1>
<p><img src="./screenshots/wallpaper.png" alt="Judge0 Widgets Wallpaper" /></p>
<p>ðŸ˜Ž Online compiler widgets for your website.</p>
<h2>About</h2>
<p><a href="https://widgets.judge0.com"><strong>Judge0 Widgets</strong></a> is a simple, but powerful platform for creating custom online compiler widgets for your website.</p>
<h2>Features</h2>
<ul>
<li>Support for <a href="https://github.com/judge0/judge0#supported-languages">60+ languages</a></li>
<li>Responsive design</li>
<li>Easy embedding with the single line of HTML</li>
<li>Light and dark theme</li>
<li>Custom and reorderable layout</li>
<li>Custom CSS and branding</li>
<li>Optional UI language</li>
<li>Custom expected output for validating userâ€™s code</li>
</ul>
<h2>Get Started</h2>
<ol>
<li>Choose a plan for <a href="https://rapidapi.com/hermanzdosilovic/api/judge0/pricing">Judge0</a> and/or <a href="https://rapidapi.com/hermanzdosilovic/api/judge0-extra/pricing">Judge0 Extra</a> (itâ€™s <strong>FREE</strong>).</li>
<li>Use your RapidAPI key at https://widgets.judge0.com.</li>
</ol>
<h2>Features for Paid Users</h2>
<p>If you are subscribed to a paid plan of at least one flavor of Judge0, you are a paid user.</p>
<p>As a paid user you can set and save <em>Custom CSS</em> or <em>Expected Output</em> for your Judge0 Widget, but as a basic user, you can just preview these features.</p>
<h2>UI Translations</h2>
<p>Do you want to have an option for translating the UI to your native language? Please contribute to this project by translating the values in <a href="./i18n/en.css">en.css</a>.</p>
<h2>FAQ</h2>
<details>
<summary>Click here to expand.</summary>
<h3>What are Judge0 and Judge0 Extra?</h3>
<p><a href="https://rapidapi.com/hermanzdosilovic/api/judge0/pricing">Judge0</a> and <a href="https://rapidapi.com/hermanzdosilovic/api/judge0-extra/pricing">Judge0 Extra</a> are two flavors of the most advanced open-source online code execution system in the world - <a href="https://github.com/judge0/judge0">Judge0</a>.</p>
<h3>What is Judge0 used for?</h3>
<p>
Judge0 is used for securely compiling and running the code your users write in the Judge0 Widgets.
</p>
<h3>Why do I need to sign up on RapidAPI?</h3>
<p>You need to sign up on RapidAPI and subscribe to <a href="https://rapidapi.com/hermanzdosilovic/api/judge0/pricing">Judge0</a> and/or <a href="https://rapidapi.com/hermanzdosilovic/api/judge0-extra/pricing">Judge0 Extra</a> to get your RapidAPI key which will be used for communicating with the Judge0.</p>
<h3>Where can I find my RapidAPI Key?</h3>
<p>
After signing up to RapidAPI, subscribe to <a href="https://rapidapi.com/hermanzdosilovic/api/judge0/pricing">Judge0</a> and/or <a href="https://rapidapi.com/hermanzdosilovic/api/judge0-extra/pricing">Judge0 Extra</a> and visit <a href="https://rapidapi.com/hermanzdosilovic/api/judge0/endpoints"><em>Endpoints</em></a>
<img src="./screenshots/rapidapi_key.png" alt="Where is my RapidAPI key?">
</p>
<h3>How do I remove Judge0 Widgets branding?</h3>
If you are a paid user you can use the <em>Custom CSS</em> feature that allows you to inject custom CSS into the widget.
<pre><code><span class="hljs-selector-id">#judge0-widgets-copyright</span> {
    <span class="hljs-attribute">display</span>: none;
}
</code></pre>
</details>
<h2>Screenshots</h2>
<p>Creating a Judge0 Widget with light theme and English UI.
<img src="./screenshots/create_widget_light_theme.png" alt="" /></p>
<p>Judge0 Widget with UI translated to Croatian.
<img src="./screenshots/widget_optional_theme_dark_hr.png" alt="" /></p>
<p>Editing Judge0 Widget settings.
<img src="./screenshots/widget_settings.png" alt="" /></p>
<p>Saving Judge0 Widget and getting instructions for embedding.
<img src="./screenshots/save_widget.png" alt="" /></p>
<p>Judge0 Widget in an online workbook with programming exercises.
<img src="./screenshots/zbirka_online.png" alt="" /></p>

